### PR TITLE
feat: root key creation from SDK

### PR DESCRIPTION
**Pull request by frdomovic**
# feat: root key creation from SDK - admin UI, calimero SDK, admin API

## Summary
In admin API -> updated root-key endpoint handler that is checks wallet meta because now ETH rootkey is also option.
Calimero SDK -> components for adding near root key and eth root key (0.0.14)
Admin UI -> added pages for logging in (adding rootkey) - index (select login option), eth login, near login.

Fixes # C20-65

## Summary by CodeRabbit

- **New Features**
	- Made specific structs and fields public for increased accessibility.
	- Introduced new structs, enums, and functions for wallet types and signature verifications.
	- Added React components for Metamask login functionality.
	- Updated context and components for Metamask and Near wallet logins.
	- Implemented a function to map Ethereum chain IDs to wallet types.

### CODE REVIEW

File: crates/server/src/admin/handlers/add_client_key.rs [hunk 1]
```diff
@@ -37,21 +37,21 @@ struct Payload {
...
 #[derive(Debug, Deserialize)]
 #[serde(rename_all = "camelCase")]
-struct SignatureMessage {
+pub struct SignatureMessage {
     nonce: String,
     application_id: String,
     timestamp: i64,
     node_signature: String,
-    message: String,
+    pub message: String,
```
Comment (frdomovic) [inline crates/server/src/admin/handlers/add_client_key.rs:45]: idk, had to add it to make it work

File: node-ui/package.json [hunk 2]
```diff
@@ -10,11 +10,13 @@
     "preview": "vite preview"
   },
   "dependencies": {
+    "@calimero-is-near/calimero-p2p-sdk": "0.0.14",
```
Comment (frdomovic) [inline node-ui/package.json:13]: will do npm push after its merged in master and this will work

File: node-ui/package.json [hunk 3]
```diff
@@ -10,11 +10,13 @@
...
     "@floating-ui/react-native": "^0.10.4",
     "@heroicons/react": "^2.1.3",
     "@near-wallet-selector/core": "^8.9.5",
     "@near-wallet-selector/here-wallet": "^8.9.5",
     "@near-wallet-selector/meteor-wallet": "^8.9.5",
+    "@near-wallet-selector/modal-ui": "^8.9.7",
```
Comment (frdomovic) [inline node-ui/package.json:19]: needed for css import as it cannot "add global css from node_modules"

File: node-ui/package.json [hunk 4]
```diff
@@ -28,7 +30,7 @@
     "buffer": "^6.0.3",
     "near-api-js": "^3.0.4",
     "prop-types": "^15.8.1",
-    "react": "^18.2.0",
+    "react": "^18.3.0",
```
Comment (frdomovic) [inline node-ui/package.json]: match sdk version, not sure how to make it compatible with other ones

File: packages/calimero-sdk/build-script-admin.sh [hunk 5]
```diff
@@ -0,0 +1 @@
+rm -rf node_modules dist lib && pnpm install && pnpm build && cd ../../node-ui/ && rm -rf node_modules && pnpm install && pnpm build
```
[the remark this answers is not in the document]
Comment (frdomovic) [reply packages/calimero-sdk/build-script-admin.sh:2]: updated

File: crates/server/src/admin/handlers/add_client_key.rs (hunk 1 above)
Comment (miraclx) [reply crates/server/src/admin/handlers/add_client_key.rs:45]: can you show the error?

File: crates/server/src/admin/service.rs [hunk 6]
```diff
@@ -178,36 +179,71 @@ async fn create_root_key_handler(
...
+                        &challenge.node_signature,
+                        recipient,
+                        &req.callback_url,
+                        &req.signature,
+                        &req.public_key,
+                    ) {
+                        let result = add_root_key(
+                            &state.store,
+                            RootKey {
+                                signing_key: req.public_key.clone(),
```
Comment (miraclx) [inline crates/server/src/admin/service.rs]:
```suggestion
                                signing_key: req.public_key,
```

this is also questionable.. a signing key is supposed to be the secret component

File: crates/server/src/admin/service.rs [hunk 7]
```diff
@@ -178,36 +179,71 @@ async fn create_root_key_handler(
...
+                            Err(e) => {
+                                error!("Failed to store root key: {}", e);
+                                (
+                                    StatusCode::INTERNAL_SERVER_ERROR,
+                                    "Failed to store root key",
+                                )
+                            }
+                        }
+                    } else {
+                        (StatusCode::BAD_REQUEST, "Invalid signature")
```
Comment (miraclx) [inline crates/server/src/admin/service.rs]: move the short-circuit closer to the conditional like you did for the ETH side

```rust
if !verifysignature::verify_near_signature(..) {
    return (StatusCode::BAD_REQUEST, "Invalid signature");
```

helps compress the indentation

File: crates/server/src/admin/service.rs [hunk 8]
```diff
@@ -178,36 +179,71 @@ async fn create_root_key_handler(
...
+                            info!("Root key added");
+                            (StatusCode::OK, "Root key added")
+                        }
+                        Err(e) => {
+                            error!("Failed to store root key: {}", e);
+                            (
+                                StatusCode::INTERNAL_SERVER_ERROR,
+                                "Failed to store root key",
+                            )
+                        }
```
Comment (miraclx) [inline crates/server/src/admin/service.rs]: also, this is duplicated, rewrite to compress

File: packages/calimero-sdk/src/wallet/MetamaskLogin/MetamaskWrapper.tsx [hunk 9]
```diff
@@ -1,14 +1,16 @@
...
 import { MetaMaskUIProvider } from "@metamask/sdk-react-ui";
 
 interface MetamaskContextProps {
   applicationId: string;
   rpcBaseUrl: string;
   successRedirect: () => void;
-  cardBackgroundColor: string | undefined;
-  metamaskTitleColor: string | undefined;
+  cardBackgroundColor?: string | undefined;
+  metamaskTitleColor?: string | undefined;
```
Comment (miraclx) [inline packages/calimero-sdk/src/wallet/MetamaskLogin/MetamaskWrapper.tsx]: this is the same, yeah?

```suggestion
  cardBackgroundColor?: string;
  metamaskTitleColor?: string;
```

File: crates/server/src/admin/handlers/add_client_key.rs (hunk 1 above)
Comment (frdomovic) [reply crates/server/src/admin/handlers/add_client_key.rs:45]: Reverted back, interesting when removed it did not show error anymore, maybe rs compiler didn't finish last time

File: crates/server/src/admin/service.rs (hunk 6 above)
Comment (frdomovic) [reply crates/server/src/admin/service.rs]: Removed clone. singing_key is public key which is used to verify signatures (?)

File: packages/calimero-sdk/src/wallet/MetamaskLogin/MetamaskWrapper.tsx (hunk 9 above)
Comment (frdomovic) [reply packages/calimero-sdk/src/wallet/MetamaskLogin/MetamaskWrapper.tsx]: Its the same, updated

File: node-ui/package.json (hunk 4 above)
Comment (MatejVukosav) [reply node-ui/package.json]: There is no difference in your change. Both will resolve to the the same version at the end

File: crates/server/src/admin/service.rs [hunk 10]
```diff
@@ -178,36 +192,47 @@ async fn create_root_key_handler(
...
+                WalletType::NEAR => {
+                    if !verifysignature::verify_near_signature(
+                        &challenge.message.nonce,
+                        &challenge.node_signature,
+                        recipient,
+                        &req.callback_url,
+                        &req.signature,
+                        &req.public_key,
+                    ) {
+                        return (StatusCode::BAD_REQUEST, "Invalid signature");
```
Comment (MatejVukosav) [inline crates/server/src/admin/service.rs]: If you want to do it properly, the root key endpoints should return "data" or "error" as the other endpoints.

File: examples/only-peers-simple/package.json [hunk 11]
```diff
@@ -10,7 +10,7 @@
     "redev": "rm -rf node_modules && pnpm install && vite"
   },
   "dependencies": {
-    "@calimero-is-near/calimero-p2p-sdk": "0.0.12",
+    "@calimero-is-near/calimero-p2p-sdk": "file:../../packages/calimero-sdk",
```
Comment (MatejVukosav) [inline examples/only-peers-simple/package.json]: I'm not too fond of this being pushed. Can u increase sdk version first and then this

File: crates/server/src/admin/handlers/add_client_key.rs [hunk 12]
```diff
@@ -37,7 +37,7 @@ struct Payload {
 
 #[derive(Debug, Deserialize)]
 #[serde(rename_all = "camelCase")]
-struct SignatureMessage {
+pub struct SignatureMessage {
```
Comment (MatejVukosav) [inline crates/server/src/admin/handlers/add_client_key.rs:40]: we have `server-primitives`. In that case, it makes more sense to put this there than make it `pub`.

File: node-ui/package.json (hunk 4 above)
Comment (frdomovic) [reply node-ui/package.json]: reverted

File: crates/server/src/admin/service.rs (hunk 7 above)
Comment (frdomovic) [reply crates/server/src/admin/service.rs]: updated

File: crates/server/src/admin/service.rs (hunk 8 above)
Comment (frdomovic) [reply crates/server/src/admin/service.rs]: updated

File: examples/only-peers-simple/package.json (hunk 11 above)
Comment (frdomovic) [reply examples/only-peers-simple/package.json]: will revert before mergning

File: crates/server/src/admin/service.rs (hunk 10 above)
Comment (frdomovic) [reply crates/server/src/admin/service.rs]: updated as rest 
Screenshot 2024-05-31 at 13 50 31

File: crates/server/src/admin/handlers/add_client_key.rs (hunk 12 above)
Comment (frdomovic) [reply crates/server/src/admin/handlers/add_client_key.rs:40]: updated

File: crates/server/src/admin/service.rs (hunk 10 above)
Comment (MatejVukosav) [reply crates/server/src/admin/service.rs]: This needs to be fixed in apps, admin and SDK at the same time

Comment (frdomovic) [reply crates/server/src/admin/service.rs]: In the sdk the code goes like if (result.error) { handle error } else { handle success } so it doesn't need any updates there or admin-ui + apps as it also uses the same sdk functions

File: node-ui/src/pages/Near.tsx [hunk 13]
```diff
@@ -0,0 +1,28 @@
+import React from "react";
+import { useNavigate } from "react-router-dom";
+import { WalletSelectorContextProvider } from "@calimero-is-near/calimero-p2p-sdk/lib/wallet/NearLogin/WalletSelectorContext";
+import NearRootKey from "@calimero-is-near/calimero-p2p-sdk/lib/wallet/NearLogin/NearRootKey";
+import ContentWrapper from "../components/login/ContentWrapper";
+
+import "@near-wallet-selector/modal-ui/styles.css";
+
+// @ts-expect-error
```
Comment (MatejVukosav) [inline node-ui/src/pages/Near.tsx:9]: remove

File: crates/server/src/admin/service.rs (hunk 6 above)
Comment (miraclx) [reply crates/server/src/admin/service.rs]: The signing key is the secret key for signing, verifying key is the public key for verifying

**Pull request merged**